bluetooth: fix format

lasers opened the pull request:
**FORMAT**
```
OLD                    |  NEW
-----------------------|-----------------------
format                 |  format 
format_prefix          |  format_device
format_no_conn         | 
format_no_conn_prefix  |
```
Fixes oops in color options.

~~Deprecates `device_separator` in favor of `string_separator`. Ya I know. Discussion first.~~
* Add _NEW! device_ <-- FUUUUU! (device_separator in bluetooth) to that list too.

**MODULE ENHANCEMENT**

I used `Microsoft Bluetooth Notebook Mouse 5000` F----- long! I fix this now...

button_toggle: mouse button to toggle visibility
* `BT: Microsoft Bluetooth Notebook Mouse 5000` --> `BT` (toggleable, icon_collapse)

always_collapse
* `BT: Microsoft Bluetooth Notebook Mouse 5000` --> `BT` (toggleable, icon_collapse)

~~**DEFAULT**~~

~~Bluetooth symbol. `font-awesome` now a hard dependency. Pretty please.~~
~~We can put `py3status` on the moon! I will put `BT:` back in after reviews...~~

~~**EDIT: EXCEPTION**~~
~~*Adds something to take care of missing `not-installed` binary.~~

tobes commented:
@lasers 

I'd like to start by thanking you for all your hard work and enthusiasm for this project.  It is really nice to have someone who is both keen, but also puts in the hard work of improving things, and for your many efforts in testing.  Both @ultrabug and myself are extremely pleased with your dedication.  We are also really keen to encourage and embrace all contributions to py3status. __Please keep it up!__

However, and __I only talk for myself from here on in__.  You need to try to take on board  some of the feedback that you are getting (admittedly mainly from myself). Do not get me wrong, it is important to challenge things (always), but when you have done this and are still getting the same response, then you need to consider why.

Personally I've spent a small part of my life doing some coding, some of it on voluntary projects like this, and some on more 'professional' things, and I'd like to hope I've made enough mistakes to have learned a few things on the way. I'll accept that I don't really know anything, because none of us do.  I'm also happy to be wrong and learn things new. Please give me the benefit of the doubt on occasion, just in case I might have some insight.

Reviewing code is hard.  It takes much more time and effort than writing code.  Writing code is easy.  Maintaining code is hard. Every line of code is a source of bugs (even comments). Every change can introduce new bugs. Every line of code adds an overhead to a project. 

I don't have the time to provide any references but I think I may have made some of these points before (sorry if I didn't)

* Do not compound several issues into one PR (especially uncontroversial/controversial stuff)

* Do not change the defaults of modules

* Do not include poorly supported unicode characters (if you do not know which are poorly supported then take take it to mean all of them)

* Try to have as few dependencies as possible

* Do not add dependencies to an already existing module unless it adds benefits which are __optional__

* Do not rename config parameters unless there is a  __clear__ reason

* Discuss things on irc (you've been invited many times)

* Try to keep conversations on issues/PRs relevant (yes I go off topic but I'm stupid)

* Do not make big changes to PRs that have been reviewed and accepted

* Change is hard, but possible

* Be patient

* Actively listen 

* font-awesome is pretty but will never be a dependency

* Accept that we can all be wrong

Also, and this is more important than the list, try to think why these items might have made it on to the list.  Why do you think I am spending my time writing this (and believe me it has not been quick)

It has got to the point when I'm not sure how best to go forward.

Once again your hard work is appreciated, but can we try to pull in the same direction a little more.

This module looks really complicated to fix correctly (I've only briefly skimmed it) so it is not the place to be trying to change the world.

@ultrabug sorry if you feel I'm not handling this very well,  talk soon.

lasers commented:
>Discuss things on irc (you've been invited many times)

... I pretty much hate IRC. Hasn't been on there for more than a year. `lasers`

Keep in mind that I'm just a hobbyist having fun. Not a programmer. Giving things a try. Improving things whenever I can. If I made mistakes, sometimes it's not intentional. Other things, I sometimes disagree... but yes... I also am trying to clean up things whenever I can.

Some codes may have been untouched for months. I'm only a set of fresh eyes and all... making changes everywhere... sometimes unnecessary.

Also, I try to help by confirming that things works... Python2and3 confirmed. Stuffs like that. I even installed stupid binaries that I don't need, but hey, all to make sure it comes out well... that's great.

>Try to keep conversations on issues/PRs relevant (yes I go off topic but I'm stupid)

I have been doing that, but yes, something things are interved. tIt's hard because Github is a social coding platform and already, I know I don't want to leave too much messages... to 23 guys.

All things you had asked of me to... I try to stay verbose... To explicitly tell you what has been changed... Comments everywhere in PRs. Some defaults are bad to begin with (like not using scrot) or too hard to shallow.

When I actually test things (binaries and all), I sometimes think I see a better way to deal with things so I try and see if I can do it... and if I pull it off with no problem, then I go ahead and make a PR. I think that's what everybody have been doing.

>font-awesome is pretty but will never be a dependency

Can't hurt to ask. Kinda like the screenshots. Introducing that one dependency could mean modules w/ sensible default options like using `{icon_bluetooth} {name}` instead of `BT: {name}`. If we have more sensible default options, then all the better... because we don't have to configure much out of the box... I think I'm rambling a bit so let's stop right there. I will still review all of my PRs and will avoid major changes on `ready for merge` tags. Sorry about that.

tobes commented:
> ... I pretty much hate IRC. 

so do I but for some things it is useful.  It might also be easier than back and forth on these PRs etc
it give others a chance to participate too

> Keep in mind that I'm just a hobbyist having fun. Not a programmer. Giving things a try. Improving things whenever I can. If I made mistakes, sometimes it's not intentional. Other things, I sometimes disagree... but yes... I also am trying to clean up things whenever I can.

Good you are keen and you take stuff on board.  We all make mistakes, I do it all the time, mistakes are good.

As I said i value your efforts, you've added a lot.

Your PRs are getting much better and you have some really good ideas like the screenshots so keep it up.

Just baby steps can sometimes be much faster than giant leaps.

And the font-awesome stuff yes it is nicer and beautiful but this project is for all users including ones with poor font support.  I have my own py3status font for my config so I know where your coming from.  Anyhow as I said somewhere there is a sneaky way to make these format changes you want including font-awesome - but I need to get down to a couple of PRs of mine in the queue before that happens (I have a lot of PRs waiting)

Remember most people only have an hour or two a week for this sort of thing - don't ask why I'm about so much...

Plus the main thing is to enjoy it.

lasers commented:
@tobes Hi. Following up on lot of things, I updated some of the PRs.

For this one, I placed `not-installed` exception outside of `post_config_hook()`. I removed `string_off` because one could construct this as a `string_error`... although I can see people wanting an option to hide bluetooth completely if no connection was made.

Based on history, I assume people will have to use something like github format trick instead.
Rough idea should be something like this... `[[\?not_zero BT: {name}]|]`

I cleaned up things a bit too. One minor issue I had so far is that sometimes bluetooth is stupid/slow... and I will see green `BT: `.... then red `BT`. The green part is a problem as I wanted to omit `: ` if there was none... but the code right now is clean. :+1:

tobes commented on py3status/modules/bluetooth.py:
```diff
@@ -3,84 +3,109 @@
+    cache_timeout: refresh interval for this module (default 10)
+    format: display format for this module (default 'BT: {format_bt}')
+    format_bt: display format for bluetooth (default '{name}')
+    string_separator: show separator only if more than one (default '|')
```
Can we just keep `device_separator` for now.  That should speed up getting things merged.

tobes commented on py3status/modules/bluetooth.py:
```diff
@@ -3,84 +3,109 @@
+
+    def post_config_hook(self):
+        self.format_bt_info = self.format_bt
+        self.format_info = self.format
```
renaming things just makes this confusing

lasers replied:
This is part of a `always_collapse` module enhancement. IRC?

tobes replied:
that can be done much more cleanly - but anyhow I think it should die - but maybe I haven't looked at the code enough - I'll do that at some point but need to find my bt dongle

tobes commented on py3status/modules/bluetooth.py:
```diff
@@ -3,84 +3,109 @@
-    format_no_conn_prefix = 'BT: '
-    format_prefix = 'BT: '
+    format = 'BT: {format_bt}'
+    format_bt = '{name}'
```
This whole format changing is quite a complicated thing.  We need to build the format based on the users config so that it works for existing users.   This all wants to be done using the format depecation functionality.

It might be easiest to do that separately from the helpers changes.

tobes commented on py3status/modules/bluetooth.py:
```diff
@@ -3,84 +3,109 @@
-    format_no_conn_prefix: prefix when there is no connected device
-        (default 'BT: ')
-    format_prefix: prefix when there is a connected device (default 'BT: ')
+    always_collapse: always collapse to hide format_bt (default False)
```
I don't like this param I think we can get things working without it.  I'd need to see the other stuff cleaned up first though.

tobes commented:
`[BT: {name}]` should be enough assuming `{name}` will be `''` or `False`

ultrabug commented:
Recent merge got this conflicting

tobes commented:
@lasers I think we should strip the collapse stuff from this PR and just concentrate on the format stuff

lasers commented:
@tobes. All done. Last thing, deprecation.

lasers commented:
@tobes Want to look at this for depreciation stuffs then we can kick it out of here?

tobes commented on py3status/modules/bluetooth.py:
```diff
@@ -55,47 +52,28 @@ class Py3status:
-    format_no_conn = 'OFF'
-    format_no_conn_prefix = 'BT: '
-    format_prefix = 'BT: '
+    format = 'BT[: {format_device}]'
```
We should build this based on the `format_prefix` in the deprecation stuff.  I think I'd not worry about `format_no_conn_prefix` as otherwise I think maybe this is not possible.

tobes replied:
Ok this is all horrible.  We want to build the `format` and `format_device` from the old config which is hard because what was `format` is now `format_device` but we can tell if `{name}` is bing used in the format that it must be an old one.

I'd still prefer this PR to be cleaner especially as we are changing the behaviour (which can't be helped)

anyhow we are moving in the right direction

lasers replied:
I'm sure I took care of `format_prefix` just now. How should I check for `{name}` or `{mac}` in format but ignore if `format_device` exists? Meh. I can test it. Could `format_no_conn` thing be done w/ the formatter?... Something like...
```
'\?if=!format_device OFF|{}'{{format_device}}'.format(config['format_prefix'])
```
I dunno what I'm doing here. Too messy for me.

tobes replied:
> `{name}` or `{mac}` in format

yeah this is tricky.  I think maybe we need some deprecation code changes to do this all properly.

`format = 'BT: [{format_device}|\?show OFF]'`  I think is pretty close to the existing output and not too ugly

lasers replied:
QQ. `format = 'BT[: {format_device}]'` is _NEW!_ sensible default to work with this _NEW!_ format along with _NEW!_ dbus.

The format I made give (new) users either red `BT` or green `BT: Device`. If we use `format = 'BT: [{format_device}|\?show OFF]'` for format, then it kinda goes against py3stautus default formats such as redgreen `DPMS`, redgreen `XSCR`, redgreen `keyboard_locks`, redgreen `module` etc.

There are few modules that acts slightly different... like `do_not_disturb` where it display redgreen `ON`/`OFF` (w/ opposite color... slightly confusing)... instead of something like redgreen `DNT`. I only mention this because if we try for sensible defaults, then people may keep their py3status config only slightly more clean.

Also with new dbus stuffs, you have to run `bluetooth` service. If not, we'll just get red `bluetooth` exception. That wasn't the case before (ie, new behavior now).

Good timing that we start showing errors in `i3bar` instead of `i3-nagbar`. So we now have red `bluetooth`, red `OFF`, and green `BT: Device`. I want to stay with red `bluetooth`, red `BT`, and green `BT: Device`.

tobes replied:
> format = 'BT[: {format_device}]' 

I can live with this

> Also with new dbus stuffs, you have to run bluetooth service. 

could you explain a little more.  This also sounds like maybe it should be a new PR once this one is done

lasers replied:
Switching from `hcitool` to `dbus` changes the behavior.

With `hcitool`, we just run the command and work with the data regardless.
With dbus, we can't obtain the data so we get exception.

I mentioned this because we're changing to `dbus` so why not fix the format at same time?

Basically, if people use Bluetooth all times, they likely have it on autostart and running all times... Otherwise, few people who manually starts bluetooth only sometimes (for audio or whatever reason), they will likely notice this red `bluetooth` instead of former `OFF` regardless.

tobes replied:
I'd have kept the hcltool support given the choice but anyhow it got merged.

Adding back is probably not worth the pain.  Anyhow we can look at that issue separately than this format stuff

tobes commented on py3status/modules/bluetooth.py:
```diff
@@ -55,47 +52,40 @@ class Py3status:
+            out = {}
+            if 'format_prefix' in config:
+                out['format'] = u'{}{{format_device}}'.format(config['format_prefix'])
+            return out
```
this is a good start.  I think the format should still be None as the default then we can check if the user has set it etc.

And we need some more checks etc

If you want I'll knock up a patch.  Won't be today as I'm off out soon

lasers replied:
Sure. I can give your patch a go only if you want to whip one up. Have fun tonight. Is there a way to `time` a module (for optimization purpose)?

tobes replied:
currently it's hard but I'd like to make it easy.

As ever I have a branch that does this somewhere in my backlog of PRs

python3 has a nice way to do it but our threading model prevents it :( anything else is a nasty hack

lasers replied:
Damn our threading model. There must be something we can use... even nasty hacks. It's better than nothing. Maybe we can use an average number out of 100 one-shots or something (excluding API modules). I don't know.

Anyway, here to say we should decide soon what to do with this. Do we mark this ready or we come up with more than just a prefix fix since this new format is much more cleaner than the old format.

lasers commented:
My bluetooth is okay after a reboot. Maybe new kernel or something. Idk. W/e.

I wanted to bring up two things I learned when performing interrogation techniques on this `bluetooth` module... and many other modules.

1) For `bluetooth`, `cache_timeout` = `10` feels somewhat sluggish only during disconnection / connection. This is likely normal for  `BT` protocol. We get used to delayed initialization / response when we come back to use `BT` after some sleep/disconnected period...... then we wait for new cache_timeout before we see a changed status.

All this delayed time adds up the response from turning on mouse to changed status. Do we want to use `5` for perceived better response time from status? Or keep it as-is? FYI: I don't use this `bluetooth` module nor this `BT` mouse much.

2) We switch to `dbus` from depreciated `hcitool` not long ago. Users now can get dbus exception instead of red `BT:OFF` / `BT` if they don't always run bluetooth like they don't always run mpd daemon (for mpd_status). The exception in question.
```
2017-03-30 01:18:12 WARNING Instance `bluetooth _anon_module_0`, user method `bluetooth` failed (DBusException) bluetooth.py line 35.
2017-03-30 01:18:12 INFO Traceback
DBusException: org.freedesktop.systemd1.NoSuchUnit: Unit dbus-org.bluez.service not found.
```
Do we... 

* Do nothing. As-is. The information is being exposed correctly. Users can dismiss it and new (same) exception will come back in seconds. May have to start running `bluetooth` daemon or something else to make it go away, but still correct. 

* `try/catch` in `post_config_hook` so we can dismiss the module away if we don't want to use.
* `try/catch` the `bluetooth` method along with `py3.error` so we can dismiss the module away if we don't want to use.

If we dismiss the module away, then obviously, we can't see the status if we start using BT headset, mouse, or whatever.

I don't know what the behaviour we should take to deal with this dbus exception... Dismiss on first chance given... or we keep running exception... without or with annoying long message... Do we give users option to dismiss it once or never? Keep it forever?

I'm annoyingly invested in making all modules very nice. Maybe I shouldn't do that. 👍

tobes commented:
@ultrabug I think this is good to go even with #862 outstanding
